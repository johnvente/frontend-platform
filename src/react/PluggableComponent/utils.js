import packageJson from '@root_path/package.json';

/**
 * Checks if a given plugin is available by attempting to dynamically import it.
 *
 * This function tries to dynamically import a plugin based on its name. It constructs
 * a path using a predefined pattern that points to the plugin's location within
 * the `node_modules` directory. If the import succeeds, it means the plugin is available.
 *
 * @param {string} pluginName - The name of the plugin to check for availability.
 * @returns {Promise<boolean>} A promise that resolves to `true` if the plugin is available, otherwise `false`.
 */
export const isPluginAvailable = async (pluginName) => {
  if (!pluginName) { return false; }

  try {
    await import(`@node_modules/@openedx-plugins/${pluginName}`);
    return true;
  } catch {
    return false;
  }
};

/**
 * Retrieves a list of plugins that match a given prefix from the project's dependencies.
 *
 * This function filters the project's dependencies listed in the `package.json` file,
 * looking for ones that match a specified prefix. The prefix is used to identify related
 * plugins within the `@openedx-plugins` namespace. Each matching plugin's name is then
 * formatted and returned as part of an array of plugin objects, each containing the
 * plugin's `id` and `name`.
 *
 * @param {string} prefix - The prefix to filter plugins by.
 * @returns {Array<Object>} An array of objects, each representing a plugin. Each object
 *                          contains the `id` and `name` of the plugin.
 */
export const getPluginsByPrefix = (prefix) => {
  const dependenciesKeys = Object.keys(packageJson.dependencies);
  return dependenciesKeys.reduce(
    (pluginsFiltered, pluginName) => {
      if (pluginName.startsWith(`@openedx-plugins/${prefix}`)) {
        const pluginFormatted = pluginName.split('/')[1];
        const pluginData = {
          id: pluginFormatted,
          name: pluginFormatted,
        };
        return [...pluginsFiltered, pluginData];
      }
      return pluginsFiltered;
    },
    [],
  );
};
