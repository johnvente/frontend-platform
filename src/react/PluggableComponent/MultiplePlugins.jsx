import React, { useState, useEffect, useRef } from 'react';
import loadable from '@loadable/component';
import useDeepCompareEffect from 'use-deep-compare-effect';
import PropTypes from 'prop-types';

import { isPluginAvailable, getPluginsByPrefix } from './utils';

const MultiplePlugins = ({
  plugins,
  pluggableComponentProps,
  prefix,
  loadingComponent,
  containerPluginsProps,
}) => {
  const [pluginComponents, setPluginComponents] = useState({});
  const loadedAllPluginsRef = useRef(null);

  useEffect(() => {
    const loadPlugins = (pluginsList) => {
      pluginsList.forEach((plugin, index) => {
        // Initially set the loading component for each plugin
        setPluginComponents(previousPluginComponents => ({
          ...previousPluginComponents,
          [plugin.id]: loadingComponent || null,
        }));

        const loadPlugin = async () => {
          try {
            const hasModuleInstalled = await isPluginAvailable(plugin.name);
            if (hasModuleInstalled) {
              const PluginComponent = loadable(() => import(`@node_modules/@openedx-plugins/${plugin.name}`));
              setPluginComponents(previousPluginComponents => ({
                ...previousPluginComponents,
                [plugin.id]: (
                  <PluginComponent {...pluggableComponentProps} />
                ),
              }));
            }
          } catch (error) {
            console.error(`Failed to load plugin ${plugin.name}:`, error);
            // Set to null in case of an error
            setPluginComponents(previousPluginComponents => ({
              ...previousPluginComponents,
              [plugin.id]: null,
            }));
          } finally {
            const isLastPlugin = index === pluginsList.length - 1;
            if (isLastPlugin) {
              loadedAllPluginsRef.current = true;
            }
          }
        };

        loadPlugin();
      });
    };

    const pluginsToLoad = prefix ? getPluginsByPrefix(prefix) : plugins;

    if (pluginsToLoad.length) {
      loadPlugins(pluginsToLoad);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useDeepCompareEffect(() => {
    const updatePluginsWithNewProps = () => {
      const updatedComponents = Object.keys(pluginComponents).reduce((previousPluginComponents, pluginKey) => {
        const PluginComponent = pluginComponents[pluginKey];
        // Check if the component is a valid React element and not a loading or error state
        if (React.isValidElement(PluginComponent)) {
          const UpdatedComponent = React.cloneElement(PluginComponent, pluggableComponentProps);
          return {
            ...previousPluginComponents,
            [pluginKey]: UpdatedComponent,
          };
        }
        return previousPluginComponents;
      }, {});

      setPluginComponents(updatedComponents);
    };

    if (loadedAllPluginsRef.current) {
      updatePluginsWithNewProps();
    }
  }, [pluggableComponentProps]);

  return (
    <div {...containerPluginsProps}>
      {Object.entries(pluginComponents).map(([pluginKey, Component]) => (
        <React.Fragment key={pluginKey}>
          {Component}
        </React.Fragment>
      ))}
    </div>
  );
};

MultiplePlugins.defaultProps = {
  plugins: [],
  pluggableComponentProps: {},
  prefix: '',
  loadingComponent: null,
  containerPluginsProps: {},
};

MultiplePlugins.propTypes = {
  plugins: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
  })),
  pluggableComponentProps: PropTypes.shape({}),
  prefix: PropTypes.string,
  loadingComponent: PropTypes.node,
  containerPluginsProps: PropTypes.shape({}),
};

export default MultiplePlugins;
