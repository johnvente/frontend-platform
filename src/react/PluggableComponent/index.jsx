import React, { useState, useEffect, useRef } from 'react';
import loadable from '@loadable/component';
import PropTypes from 'prop-types';
import useDeepCompareEffect from 'use-deep-compare-effect';

import { isPluginAvailable } from './utils';
import MultiplePlugins from './MultiplePlugins';

/**
 * PluggableComponent - A component that allows dynamic loading and replacement of child components.
 *
 * @param {object} props - Component props
 * @param {React.ReactNode} props.children - Child elements to be passed to the plugin component
 * @param {React.ReactNode} props.loadingComponent - Component to be rendered while the plugin is loading
 * @param {string} props.as - String indicating the module to import dynamically
 * @param {string} props.id - Identifier for the plugin
 * @param {object} props.pluggableComponentProps - Additional props to be passed to the dynamically loaded component
 * @param {string} props.pluginsPrefix - Prefix used to identify plugins. This is used without the 'plugins' prop.
 * @param {Array<Object>} props.plugins -
  *  An array of plugin configurations.
  *  Each configuration is an object that may include plugin-specific properties.
 * @param {object} props.containerPluginsProps -
  * Props to be spread on the container that wraps multiple plugin components.
  * Useful for passing classNames or styles for layout.
 * @returns {React.ReactNode} - Rendered component
 */
const PluggableComponent = ({
  children,
  loadingComponent,
  as,
  id,
  pluginsPrefix,
  plugins,
  containerPluginsProps,
  ...pluggableComponentProps
}) => {
  const [newComponent, setNewComponent] = useState(children || null);
  const loadedComponentRef = useRef(null);
  const [isLoadingComponent, setIsLoadingComponent] = useState(false);
  const hasConfigForMultiplePlugins = pluginsPrefix || plugins.length;

  useEffect(() => {
    const loadPluginComponent = async () => {
      setIsLoadingComponent(true);

      try {
        const hasModuleInstalled = await isPluginAvailable(as);

        if (hasModuleInstalled) {
          const PluginComponent = loadable(() => import(`@node_modules/@openedx-plugins/${as}`));

          const component = children ? (
            <PluginComponent key={id} {...pluggableComponentProps}>
              {children}
            </PluginComponent>
          ) : (
            <PluginComponent key={id} {...pluggableComponentProps} />
          );

          setNewComponent(component);
          loadedComponentRef.current = true;
        }
      } catch (error) {
        console.error(`Failed to load plugin ${as}:`, error);
      } finally {
        setIsLoadingComponent(false);
      }
    };

    const hasNoPlugins = !pluginsPrefix && !plugins.length;

    if (hasNoPlugins) {
      loadPluginComponent();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [id, as]);

  useEffect(() => {
    if (newComponent && children && loadedComponentRef.current) {
      const updatedComponent = React.cloneElement(newComponent, pluggableComponentProps, children);
      setNewComponent(updatedComponent);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [children]);

  useDeepCompareEffect(() => {
    if (newComponent && loadedComponentRef.current) {
      const updatedComponent = React.cloneElement(newComponent, pluggableComponentProps, children);
      setNewComponent(updatedComponent);
    }
  }, [pluggableComponentProps]);

  if (hasConfigForMultiplePlugins) {
    return (
      <MultiplePlugins
        prefix={pluginsPrefix}
        {...{
          plugins,
          pluggableComponentProps,
          loadingComponent,
          containerPluginsProps,
        }}
      />
    );
  }

  if (isLoadingComponent && loadingComponent) {
    return loadingComponent;
  }

  return newComponent;
};

PluggableComponent.defaultProps = {
  loadingComponent: null,
  plugins: [],
  pluginsPrefix: '',
  containerPluginsProps: {},
  children: undefined,
  as: '',
  id: '',
};

PluggableComponent.propTypes = {
  children: PropTypes.node,
  loadingComponent: PropTypes.node,
  as: PropTypes.string,
  id: PropTypes.string,
  pluginsPrefix: PropTypes.string,
  containerPluginsProps: PropTypes.shape({}),
  plugins: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
  })),
};

export default PluggableComponent;
