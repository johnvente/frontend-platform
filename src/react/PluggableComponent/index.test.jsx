import React, { useState } from 'react';
import {
  render, waitFor, screen, fireEvent,
} from '@testing-library/react';
import PluggableComponent from '.';

const ToggleContentComponent = () => {
  const [showContent, setShowContent] = useState(false);

  return (
    <div>
      <button type="button" onClick={() => setShowContent((prev) => !prev)}>
        Toggle Content
      </button>
      {showContent && <div data-testid="toggle-content">Toggle On</div>}
    </div>
  );
};

describe('PluggableComponent', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  test('renders correctly', async () => {
    const handleClickMock = jest.fn();
    const props = {
      title: 'button title',
      handleClick: handleClickMock,
    };

    const { container } = render(
      <PluggableComponent
        id="pluggableComponent"
        as="any-mfe-plugins-test"
        {...props}
      >
        <h1>Hi this is the original component</h1>
      </PluggableComponent>,
    );

    await waitFor(() => {
      const buttonComponent = screen.getByTestId('button-test');
      expect(buttonComponent).toBeInTheDocument();
      expect(screen.getByText(props.title)).toBeInTheDocument();
      fireEvent.click(buttonComponent);
      expect(handleClickMock).toHaveBeenCalled();
      expect(container).toMatchSnapshot();
    });
  });

  test('loads children component when import is invalid', async () => {
    render(
      <PluggableComponent id="est-pluggable" as="invalid import">
        <div data-testid="plugin">Plugin Loaded</div>
      </PluggableComponent>,
    );

    await waitFor(() => {
      const defaultComponent = screen.getByTestId('plugin');
      expect(screen.getByText('Plugin Loaded')).toBeInTheDocument();
      expect(defaultComponent).toBeInTheDocument();
    });
  });

  test('loads children component when import is empty', async () => {
    render(
      <PluggableComponent
        id="test-pluggable"
        as=""
      >
        <div data-testid="plugin">Plugin Loaded</div>
      </PluggableComponent>,
    );

    await waitFor(() => {
      const defaultComponent = screen.getByTestId('plugin');
      expect(screen.getByText('Plugin Loaded')).toBeInTheDocument();
      expect(defaultComponent).toBeInTheDocument();
    });
  });

  test('returns null when do not have children and import is invalid', async () => {
    render(
      <PluggableComponent
        id="test-pluggable"
        as="invalid-module"
      />,
    );

    await waitFor(() => {
      expect(screen.queryByTestId('plugin')).not.toBeInTheDocument();
    });
  });

  test('updates component when props change', async () => {
    const { rerender } = render(
      <PluggableComponent
        id="test-pluggable"
        as="any-mfe-plugins-test"
        title="Testing title component"
      />,
    );

    await waitFor(() => {
      expect(screen.getByText('Testing title component')).toBeInTheDocument();
    });

    rerender(
      <PluggableComponent
        id="test-pluggable"
        as="any-mfe-plugins-test"
        title="Testing a new title component"
      />,
    );

    await waitFor(() => {
      expect(screen.getByText('Testing a new title component')).toBeInTheDocument();
    });
  }); 

  test('updates component when children change', async () => {
    const { getByText, getByTestId } = render(
      <PluggableComponent
        id="test-pluggable"
        as="default-children"
      >
        <ToggleContentComponent />
      </PluggableComponent>,
    );

    await waitFor(() => {
      const toggleContent = screen.queryByTestId('toggle-content');
      expect(toggleContent).not.toBeInTheDocument();
    });

    const toggleButton = getByText('Toggle Content');
    fireEvent.click(toggleButton);

    await waitFor(() => {
      const toggleContent = getByTestId('toggle-content');
      expect(toggleContent).toBeInTheDocument();
      expect(toggleContent).toHaveTextContent('Toggle On');
    });
  });

  test('renders loadingComponent while the plugin is loading', async () => {
    jest.mock('./utils', () => ({
      isPluginAvailable: jest.fn().mockImplementation(
        () => new Promise((resolve) => {
          setTimeout(() => {
            resolve(true);
          }, 1000);
        }),
      ),
    }));

    await waitFor(() => {
      const { getByText } = render(
        <PluggableComponent
          id="test-pluggable"
          as="any-mfe-plugins-test"
          title="Test Pluggable"
          loadingComponent={<div>Loading...</div>}
        />,
      );
      expect(getByText('Loading...')).toBeInTheDocument();
    });
  });

  test('renders multiple plugins', async () => {
    const MockPluginComponent = () => <div data-testid="plugin-test">Mocked Plugin Component</div>;

    // Mock the dynamic import to resolve with the MockPluginComponent
    jest.mock('@node_modules/@openedx-plugins/any-mfe-plugins-test', () => MockPluginComponent, { virtual: true });

    const { getByTestId } = render(
      <PluggableComponent
        id="test-pluggable"
        as=""
        plugins={[{ id: 'plugin-test-id', name: 'any-mfe-plugins-test' }]}
      />,
    );

    await waitFor(() => {
      const pluginComponent = getByTestId('plugin-test');
      expect(pluginComponent).toBeInTheDocument();
      expect(pluginComponent).toHaveTextContent('Mocked Plugin Component');
    });
  });

  test('renders multiple plugins with prefix', async () => {
    const MockPluginComponent = () => <div data-testid="plugin-test">Mocked Plugin Component</div>;

    // Mock the dynamic import to resolve with the MockPluginComponent
    jest.mock('@node_modules/@openedx-plugins/any-mfe-plugins-test', () => MockPluginComponent, { virtual: true });

    const { getByTestId } = render(
      <PluggableComponent
        id="test-pluggable"
        as=""
        pluginsPrefix="any-mfe-plugins"
        plugins={[]}
      />,
    );

    await waitFor(() => {
      const pluginComponent = getByTestId('plugin-test');
      expect(pluginComponent).toBeInTheDocument();
      expect(pluginComponent).toHaveTextContent('Mocked Plugin Component');
    });
  });
});
