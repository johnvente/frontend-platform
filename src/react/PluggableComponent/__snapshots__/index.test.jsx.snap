// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluggableComponent renders correctly 1`] = `
<div>
  <button
    data-testid="button-test"
    type="button"
  >
    button title
  </button>
</div>
`;
